subsys/net/lib/sockets/sockets_can.c

```diff
@@ -82,32 +93,82 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
 			     union net_proto_header *proto_hdr,
 			     int status, void *user_data)
 {
-	NET_DBG("ctx %p pkt %p st %d ud %p", ctx, pkt, status, user_data);
+	/* The ctx parameter is not really relevant here. It refers to first
+	 * net_context that was used when registering CAN socket.
+	 * In practice there can be multiple sockets that are interested in
+	 * same CAN id packets. That is why we need to implement the dispatcher
+	 * which will give the packet to correct net_context(s).
+	 */
+	struct net_pkt *clone = NULL;
+	int i;
 
-	/* if pkt is NULL, EOF */
-	if (!pkt) {
-		struct net_pkt *last_pkt = k_fifo_peek_tail(&ctx->recv_q);
+	for (i = 0; i < ARRAY_SIZE(receivers); i++) {
+		struct zcan_frame *zframe = (struct zcan_frame *)
+							pkt->buffer->data;
+		struct can_frame frame;
 
-		if (!last_pkt) {
-			/* If there're no packets in the queue, recv() may
-			 * be blocked waiting on it to become non-empty,
-			 * so cancel that wait.
+		if (receivers[i].iface != net_pkt_iface(pkt)) {
+			continue;
+		}
+
+		can_copy_zframe_to_frame(zframe, &frame);
+
+		if ((frame.can_id & receivers[i].can_mask) !=
+		    (receivers[i].can_id & receivers[i].can_mask)) {
+			continue;
+		}
+
+		if (pkt != NULL && ARRAY_SIZE(receivers) > 1) {
```

Comment:
Don't you need to check for i=0 instead of ARRAY_SIZE(receivers) > 1 here?
Because the original pkt must can be passed so it is unrfed somewhere.

Reply:
> Don't you need to check for i != 0 instead of ARRAY_SIZE(receivers) > 1 here?

In order to simplify pkt housekeeping, I just cloned the packet if there are multiple receiver. After revisiting this one after some time, it looks quite non-optimal. I will rework this a bit more.

Reply:
Still I would keep the original pkt and only clone if i > 0.
But this works too.